ocaml/xapi/stream_vdi.ml

```diff
@@ -100,6 +100,53 @@ let write_block ~__context filename buffer ofd len =
     then raise (Api_errors.Server_error (Api_errors.client_error, [ExnHelper.string_of_exn e]))
     else raise e
 
+(* Example = [{"length": 2097152, "flags": 0}, {"length": 1902848, "flags": 3}]*)
+type descriptor = {
+  length : int64;
+  status_flags : int32;
+}
+
+let cycle_descriptors descriptor_list offset =
+  let rec range a b =
+    if a > b then []
+    else a :: range (a + 1) b
+  in
+  (* Insert proper not_empty function here *)
+  let not_empty status_flag =
+    match (Int32.to_int status_flag) with
+      | 3 -> true
+      | _ -> false
+  in
+  let find_blocks descriptor offset increment =
+    match (not_empty descriptor.status_flags) with
+        | true -> begin
+            let start_chunk = offset / increment in
+            (* If the chunk ends at the boundary don't include the next chunk *)
+            let end_chunk = (Int64.to_int descriptor.length + offset - 1) / increment in
+            let chunks = range start_chunk end_chunk in
+            chunks, descriptor.length
+            end
+        | false -> [], descriptor.length
+  in
+  let check_exists a b =
+    if List.exists ((=)a) b = false then b@[a]
+       else b
+  in
+
+  let rec add_unique_chunks acc = function
+    | [] -> acc
+    | x::xs -> add_unique_chunks (check_exists x acc) xs
+  in
+
+  let rec process acc offset = function
+    | [] -> acc, offset
+    | x::xs -> begin
+        let chunks, add_offset = find_blocks x offset (Int64.to_int chunk_size) in
+        process (add_unique_chunks acc chunks) (offset + (Int64.to_int add_offset)) xs
+    end
+  in
+  let chunks, offset = process [] offset descriptor_list  in
+  List.map (fun a -> a * Int64.to_int chunk_size) chunks, offset
```

Comment:
@stefanopanella Updated to return chunks in chunk_size format

Reply:
Discussed that this is not actually needed here. Can do in Stefano's function instead